core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -379,6 +379,41 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     }
   executorAllocationManager.foreach(_.start())
 
+  // Use reflection to instantiate listeners specified via the `spark.extraListeners` configuration
+  // or the SPARK_EXTRA_LISTENERS environment variable
+  try {
+    val listenerClassNames: Seq[String] = {
+      val fromSparkConf = conf.get("spark.extraListeners", "").split(',')
+      val fromEnvVar = Option(conf.getenv("SPARK_EXTRA_LISTENERS")).getOrElse("").split(',')
+      (fromSparkConf ++ fromEnvVar).map(_.trim).filter(_ != "")
+    }
+    for (className <- listenerClassNames) {
+      val listenerClass = Class.forName(className).asInstanceOf[Class[_ <: SparkListener]]
+      val listener = try {
+        listenerClass.getConstructor(classOf[SparkConf]).newInstance(conf)
+      } catch {
```

Comment:
So using exceptions for basically branching is not really great. You can alternatively use:
http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getDeclaredConstructors%28%29 to find the right constructor.